### PR TITLE
Fix customer balance n+1

### PR DESCRIPTION
**Pull request by sauloperez**
#### What? Why?

This solves [v3.3.1](https://github.com/openfoodfoundation/openfoodnetwork/releases/tag/v3.3.1)'s performance regression when calculating the customer balance.

It's simpler and many orders of magnitude more efficient to ask the DB to aggregate the customer balance based on their orders. It removes a nasty N+1. To do that I created a query object based on the work started in https://github.com/openfoodfoundation/openfoodnetwork/pull/5031. 

This is a first attempt at solving the negative performance impact of our current approach but the best solution in the long term is to either offload the balance calculation to a background job to be performed overnight so we don't go through all customer orders every time. No doubt it'll be slow for hubs that have thousands of customers with thousands of orders.

This doesn't `close` https://github.com/openfoodfoundation/openfoodnetwork/issues/4732. That needs small and fast iterations using this new approach to each of the pages displaying the customer balance.

#### What should we test?

This involves two sets of tests. 

On one hand, we need to smoke test customer balance in /admin/customers. Just to confirm that, as the automated tests tell, we haven't changed a bit of its behavior.

On the other, we need to check how the new implementation works. For that, the user you test it with needs to have the feature enabled for them. To do that, we need to add the account's email address to the list of beta users. Then, with a shopper account create some orders and payments and check:

* As Hub manager make adjustments to the orders and payments. Including voiding payments and orders. Add multiple payments to an order and include addition payments so that the orders also have more payments than the order amount.
* Cancel an order with a payment.
* Void a payment on an active order
* As you make the changes in step 2 check the Shopper balance in /admin/customers

The figures should match with all the changes you make. It'd be good to go through all of the states of the order. If there's one broken, please, don't let this pass without a test covering that case.

#### Release notes

Remove N+1 when fetching customers' balance from /admin/customers and fix `payment_total` for void payments.
Changelog Category: Technical changes

### CODE REVIEW

**Comment by Matt-Yorkley** [comment]
There's another issue on this page, which is lack of pagination. The "load more" buttons on the page don't make new requests, all the data for all customers is loaded in Angular when the page loads...

**Comment by Matt-Yorkley** [comment]
I did some quick comparisons with master and this branch:

Screenshot from 2020-11-11 17-24-24
Screenshot from 2020-11-11 17-31-11

Nice :100:

**Review by Matt-Yorkley — APPROVED** [review]

**Review by luisramos0 — COMMENTED** [review]
Did you consider this PR?
https://github.com/openfoodfoundation/openfoodnetwork/pull/5031/files
It must include payments.sum(amount). order.payment_total is not the same. I recommend you look at prod data and see the difference for completed orders and for cancelled orders.

**Comment by sauloperez** [comment]
Yes, I looked at it I didn't see why. Is it because `payment_total` is what the customer needs to pay rather than what was already paid?

**Comment by lin-d-hop** [comment]
I believe because an order can have multiple payments but spree_orders.payments only records the first payment on an order

**Comment by luisramos0** [comment]
yes, I checked the code and the DBs at the time I reviewed the PR and then afterwards. It is my opinion that the algorithm is 100% correct in the pending PR and we have to follow it in this new query.

**Comment by luisramos0** [comment]
or maybe I should say: please prove me wrong here, I may be wrong, this is a tricky space. But I think you need to do that with examples from prod data, ok? This is for the inclusion of payments vs payment_total in the query.

anyway, I am 100% sure you need to include payment_total (wherever it is coming) from cancelled orders but not those ordfers.total, following this logic:
https://github.com/openfoodfoundation/openfoodnetwork/pull/5031/files#diff-87a07ead94986a81b67425065aaa01286682292f4612ada1b384a91e8ddf1d3bR418

File: app/controllers/admin/customers_controller.rb
Comment (Matt-Yorkley) [inline app/controllers/admin/customers_controller.rb:71]: So can we just change this line from `.not_state(:cancelled)` to `.where(state: ['complete', 'canceled'])` and get the same result as the pending PR?

Comment (Matt-Yorkley) [reply app/controllers/admin/customers_controller.rb:71]: Ah no, I see. We need to sum all payments on both cancelled and complete orders and sum all totals on complete orders. I'm sure we can manage this in a single query :+1:

**Comment by sauloperez** [comment]
Thanks everyone! It was clear that had to validate my ideas early on with you.

Yes, I think that logic is right @luisramos0 . For the record, I also found out that `payment_total` is a denormalized column that we populate when the payments are successfully processed in https://github.com/openfoodfoundation/openfoodnetwork/blob/8c9960854f03d485df730da663bfbbeb38ebe319/app/models/spree/order.rb#L502-L517 so we can rely on it.

**Comment by sauloperez** [comment]
This last commit should implement that same logic but in SQL @luisramos0

**Comment by luisramos0** [comment]
I think you are missing the fact that the join is by email not customerid. Guest orders will not have customer id.

**Comment by Matt-Yorkley** [comment]
Red build... looks like it needs a few tweaks.

**Comment by sauloperez** [comment]
Yes, that's why it is still draft. It's not yet done.

**Comment by sauloperez** [comment]
> I think you are missing the fact that the join is by email not customerid. Guest orders will not have customer id.

that's right, but then, there's no customer for whom to show the balance, right? oh wait, we do have a "phantom" customer created every time they buy. Got it.

**Comment by luisramos0** [comment]
I was thinking just about customer_id null in spree_orders, guest users,
have a look in prod:
select o.email, o.customer_id, o.user_id from spree_orders o where o.state
= 'complete' and o.customer_id is null and exists (select 1 from customers
where email = o.email);

But now that you mention, multiple customer records with same email:
select email, count(*) from customers group by email having count(*) > 1;

I think this should all be grouped by email so the balance is real.

**Comment by sauloperez** [comment]
> I think this should all be grouped by email so the balance is real.

That's exactly what I thought. Since they are identified by email grouping by that will do the trick.

**Comment by sauloperez** [comment]
For the record (still to be confirmed querying production data): in a guest order, there's no customer_id provided by the request. In `Spree::Order#associate_customer` we look up any customer record with the order email and associate it with the order. If none was found, in `Spree::Order#ensure_customer` we create one with the provided email. In upcoming checkouts, it'll be associated with the new order due in `#associate_customer`. The resulting order has an email and a customer_id.

File: app/controllers/admin/customers_controller.rb
Comment (sauloperez) [inline app/controllers/admin/customers_controller.rb:21]: The serializer's class name is to me a clear indication that we should extract an API endpoint out of this, together with the if branch of #collection.

**Comment by sauloperez** [comment]
I'll get to the bottom of all this, I swear :sweat_smile:

**Comment by Matt-Yorkley** [comment]
Awesome :muscle:

Two thoughts;
- it might be nice to do some before and after performance analysis with production data again (see my earlier comment) to get an idea of the impact. I can do it if you like, it takes ~5 minutes.
- do we need to do anything with `user_balance_calculator.rb`? Is it still needed elsewhere? Do we want to do anything with the changes in #5031?

**Comment by lin-d-hop** [comment]
I believe user_balance_calculator is used in the Payment Methods report. 
At some point it would be nice to clean all the user balances up but I'm thinking this will make a good first step in 'Users can pay with credits and balances' work .... when we get to that.

**Comment by sauloperez** [comment]
Re 

> it might be nice to do some before and after performance analysis with production data again (see my earlier comment) to get an idea of the impact. I can do it if you like, it takes ~5 minutes.

It's exactly what I was thinking. If you could just do what you did last time so we can compare :pray: 

Re the UserBalanceCalculator, it's out of the scope of this PR as it's used in at least 2 more places: the report @lin-d-hop mentions and account balance in the shopfront (there might be others). If things go as planned we can iteratively reuse the new implementation there as we fix bugs.

**Comment by luisramos0** [comment]
Hello. Re excluding cart orders. Id check if cart orders can or not have
completed payments...

I dont think we should move this pr forward without the other one that
improves user_balance_calculator.rb the numbers must match across the
platform #productconsistency

**Comment by lin-d-hop** [comment]
I agree with @luis because this is a real pain point for users that rely on credits to know when to invoice customers. However I also see that this is a conversation for product curation because it is a reasonable sized piece of work to actually create product consistency here. This is a pain point I'd love to resolve.

One of the challenges is that the balance is actually used in different ways:
In reports it *should* show a running balance over time.
In the shopfront and Customer's page it only needs to show a current balance.

**Comment by sauloperez** [comment]
> Re excluding cart orders. Id check if cart orders can or not have completed payments...

I can't manage to create a payment for a cart order from the backoffice. As for the frontoffice, I have no idea how to do it unless I do a checkout.

My concern about touching all other pages using customer balance is that the scope no longer is a bugfix but yes, giving up on consistency it's not good. I'm happy to move it forward now that I have a good understanding if we decide to go on.

**Comment by RachL** [comment]
> If servers hold it so far, I hope they can do so a bit longer.

Server is holding on but FYI this is one of our top support request currently. Even with hiding the column by default, big users that rely on tagging to manage their orders are struggling to use this page regularly. If it is useful, you can test with hub 922 with FR prod data.

**Comment by sauloperez** [comment]
:hankey: 

> If it is useful, you can test with hub 922 with FR prod data.

this is gold! thanks @RachL ! Then, it seems there's no way around getting this consistent across pages.

**Comment by luisramos0** [comment]
I just meant we can verify the other pr is compatible with this one and
correct, and just move it forward now as well.

File: app/controllers/admin/customers_controller.rb
Comment (mkllnk) [inline app/controllers/admin/customers_controller.rb]: I think that this is correct as a refactor and even though it's maybe not the right value in all cases, it should be the same as before. That is something we can easily test with production data though.

File: app/services/customers_with_balance.rb
Comment (mkllnk) [inline app/services/customers_with_balance.rb:35]: What about other states? `address`, `delivery` and `payment` all happen before the checkout is complete.

**Review by mkllnk — CHANGES_REQUESTED** [review]
Nice work! :+1:

I think that this is almost ready for testing and that we should work further on this. It can't be that our code is so brittle around this. Cached values are always a nightmare.

I almost approved this but I saw that there is a failing test related to the code changes. So that needs fixing first.

The commit *Update order payment_total when voiding a payment* contains only spec changes. Did you revert that part? I guess if we would update `payment_total` then we would need to run a migration to update existing data as well. It's a cached value after all and should equal the sum of all valid payments associated to that order. That might be something to check periodically and report to Bugsnag if it isn't true.

**Comment by lin-d-hop** [comment]
I just had a look at this. It looks like a good solution. Thumbs up to Maikel's states catches.

I battered my head against balances for a while so I'm actually really glad to see that it it taking all dev heads to get to a good solution. Turns out it IS a really hard problem in OFN. Certainly one for the tech debt pile to work on consistency throughout the app asap.

@lbwright22 has done some excellent testing on balances so perhaps she might help with the testing of this @filipefurtad0 ...?

**Comment by sauloperez** [comment]
> I just meant we can verify the other pr is compatible with this one and
correct, and just move it forward now as well.

Why would we do so if we know it's an approach that hampers performance?

Yes, @mkllnk I realized that commit message was misleading. It turned out that the order model was already properly updating the `payment_total` column through the order updater. I'll need to rephrase it and push force.

Re these other states. You are totally right. I didn't initially work on them because this was meant as a rather quick fix but now that this seems to become a proper refactor they need to be looked at.

@lin-d-hop does that mean I should move it forward? I'm waiting for the product's team signal.

**Comment by sauloperez** [comment]
I opened https://github.com/openfoodfoundation/openfoodnetwork/pull/6516 to enable toggling this new customer balance implementation to only ourselves. This will enable us to ship a PR per page/area without ending up with inconsistent balances in production. Otherwise, fixing all balances in this PR will be unreviewable.

File: app/controllers/admin/customers_controller.rb
Comment (sauloperez) [inline app/controllers/admin/customers_controller.rb:32]: This and the `else` branch at `#customers_relation` will be removed once we release it to all users. Now it's extra complexity we trade in favor of the flexibility the feature toggle brings.

File: app/services/customers_with_balance.rb
Comment (sauloperez) [reply app/services/customers_with_balance.rb]: Solved in 65b33b931 @mkllnk . Thanks for the heads-up

File: app/controllers/admin/customers_controller.rb
Comment (sauloperez) [reply app/controllers/admin/customers_controller.rb]: For the record:

Screenshot from 2020-12-16 12-52-21

Comment (sauloperez) [reply app/controllers/admin/customers_controller.rb]: Confirmed. There's no order in state `returned` in any of the triumvirate instances.

Comment (sauloperez) [reply app/controllers/admin/customers_controller.rb]: alright! all states covered! `returned` however might need more work in the future because return authorizations don't cause a partial refund AFAIK so we can't adjust the customer balance accordingly. That's the best we can do without touching that other feature.

**Comment by sauloperez** [comment]
alright! all states covered by code and tests! `returned` however might need more work in the future because return authorizations don't cause a partial refund AFAIK so we can't adjust the customer balance accordingly. That's the best we can do without touching that other feature.

**Review by mkllnk — APPROVED** [review]
Great work, Pau!

This is a tricky area and I think that both @Matt-Yorkley and @andrewpbrett should review this before we test and merge it.

I wrote a script (see below) to generate all JSON for all customers and ran it with AU production data. First the timings:

Old master code: 4m27s
This PR with old method: 4m9s. :shrug: 
This PR with new method: 1m12s :heavy_check_mark: 

I compared the JSON result and the new PR produces exactly the same code with the feature toggle off. :heavy_check_mark: 

The result changes though when switching the feature toggle on. That's expected because we now consider the value of cancelled orders. It affects 1,100 out of 16,988 customers, 6.5%. In general, it seems to bring the customer balances closer to zero. Some customers now have a zero balance while they were negative before. I could theorise how this could happen but maybe we should look at some cases to judge if the new value makes sense. @lin-d-hop may have some good examples of expected balances which we could verify.

I guess that this is a change we need to communicate before we toggle the feature. My scripts below may be good to find the affected customers.

Since this is feature-toggled and the result is the same with the feature off, I'll approve it. :heavy_check_mark: 



> This is a first attempt at solving the negative performance impact of our current approach but the best solution in the long term is to either offload the balance calculation to a background job to be performed overnight so we don't go through all customer orders every time. No doubt it'll be slow for hubs that have thousands of customers with thousands of orders.

Maybe a bit off-topic here but this needs more discussion. You are basically talking about introducing another cached value. I would be tempted to update the cached value whenever the order changes to avoid it being out of sync.

Test script for master:

```rb
#!/usr/bin/env ruby

# time ./bin/rails r script/export-customers.rb > customers-json-master
# 4m26.604s

require "open_food_network/user_balance_calculator"

def customer_tags_by_id(enterprise)
  customer_tags = ::ActsAsTaggableOn::Tag.
    joins(:taggings).
    includes(:taggings).
    where(taggings:
            { taggable_type: 'Customer',
              taggable_id: Customer.of(enterprise.id),
              context: 'tags' })

  customer_tags.each_with_object({}) do |tag, indexed_hash|
    tag.taggings.each do |tagging|
      customer_id = tagging.taggable_id
      indexed_hash[customer_id] ||= []
      indexed_hash[customer_id] << tag.name
    end
  end
end

Enterprise.order(:id).find_each do |enterprise|
  tag_rule_mapping = TagRule.mapping_for(Enterprise.where(id: enterprise.id))
  customer_tags = customer_tags_by_id(enterprise)

  Customer.of(enterprise.id).order(:id).
    includes(:bill_address, :ship_address, user: :credit_cards).
    find_each do |customer|

    serializer = Api::Admin::CustomerSerializer.new(
    #serializer = ::Api::Admin::CustomerWithCalculatedBalanceSerializer.new(
      customer,
      tag_rule_mapping: tag_rule_mapping,
      customer_tags: customer_tags,
    )

    puts serializer.to_json
  end
end
```

Test script for the new balance:
```rb
#!/usr/bin/env ruby

# time ./bin/rails r script/export-customers-new.rb > customers-json-new
#
# feature off:
#   4m8.858s
#
# feature on:
#   1m12.232s

require "open_food_network/user_balance_calculator"

def customer_tags_by_id(enterprise)
  customer_tags = ::ActsAsTaggableOn::Tag.
    joins(:taggings).
    includes(:taggings).
    where(taggings:
            { taggable_type: 'Customer',
              taggable_id: Customer.of(enterprise.id),
              context: 'tags' })

  customer_tags.each_with_object({}) do |tag, indexed_hash|
    tag.taggings.each do |tagging|
      customer_id = tagging.taggable_id
      indexed_hash[customer_id] ||= []
      indexed_hash[customer_id] << tag.name
    end
  end
end

Enterprise.order(:id).find_each do |enterprise|
  tag_rule_mapping = TagRule.mapping_for(Enterprise.where(id: enterprise.id))
  customer_tags = customer_tags_by_id(enterprise)

  CustomersWithBalance.new(enterprise.id).query.order(:id).
    includes(:bill_address, :ship_address, user: :credit_cards).
    find_each do |customer|

    serializer = ::Api::Admin::CustomerWithBalanceSerializer.new(
      customer,
      tag_rule_mapping: tag_rule_mapping,
      customer_tags: customer_tags,
    )

    puts serializer.to_json
  end
end
```

**Comment by sauloperez** [comment]
wow, @mkllnk that's some serious review :muscle: !

> In general, it seems to bring the customer balances closer to zero. Some customers now have a zero balance while they were negative before.

And I think that's good. We're considering all possible order states and I'd say this makes the feature actually useful. I agree with you, it needs some product thinking now.

> I would be tempted to update the cached value whenever the order changes to avoid it being out of sync.

yes, I also thought about that. I was just throwing the idea because balances will hardly be solved without offloading. We can also offload the calculation but right after the purchase. We'll see. We'll probably do well without it for some more time.

File: app/serializers/api/admin/customer_with_balance_serializer.rb
Comment (andrewpbrett) [inline app/serializers/api/admin/customer_with_balance_serializer.rb:8]: This is just for my own understanding: We're delegating to the `:object` here, which is an instance of `Customer` - but the method `balance_value` isn't defined in the Customer class. So where does this delegation get `balance_value` from?

**Review by andrewpbrett — APPROVED** [review]
This looks great, awesome work @sauloperez! 

I only did some basic checks with placing an order and checking the new customer balances to see that they made sense, since @mkllnk already did an amazing job coming up with a benchmarking procedure. 

The two comments are only for my understanding. I think this is good to go, especially since it's behind a feature toggle for now.

File: app/controllers/admin/customers_controller.rb
Comment (sauloperez) [reply app/controllers/admin/customers_controller.rb:79]: yes, this is a trick to make it possible to chain with further AR relations but Rails 4 introduced [none](https://apidock.com/rails/v4.0.2/ActiveRecord/QueryMethods/none) to do that and can probably replace it. Note this is existing code though.

File: app/serializers/api/admin/customer_with_balance_serializer.rb
Comment (sauloperez) [reply app/serializers/api/admin/customer_with_balance_serializer.rb:8]: This serializer relies on duck typing so `:object` must respond to `balance_value`, which is what the objects returned by `CustomersWithBalance#query` do. I should probably document this above line 8.

Comment (andrewpbrett) [reply app/serializers/api/admin/customer_with_balance_serializer.rb:8]: Yeah I don't think it needs to be changed before merging, but it's a little hard to understand. Even knowing what you just said, and looking at `CustomersWithBalance`, the `query` method returns a scope on `Customer`, so I'm still not sure how we actually calculate the balance on a Customer; is it the `AS balance_value` line in `CustomersWithBalance` that's doing the heavy lifting?

I'm asking these questions knowing that I could probably search around and figure out exactly what's going on; but my point is that we could potentially make this part easier to understand without making someone do that (like the next person who comes along and looks at this, who will probably be one of us in six months 😁)

File: app/controllers/admin/customers_controller.rb
Comment (andrewpbrett) [reply app/controllers/admin/customers_controller.rb:79]: Ah yeah `none` would make it more obvious what's happening.

File: app/serializers/api/admin/customer_with_balance_serializer.rb
Comment (sauloperez) [reply app/serializers/api/admin/customer_with_balance_serializer.rb:8]: Fair enough. I'll try to doc it in the class itself.

Comment (mkllnk) [reply app/serializers/api/admin/customer_with_balance_serializer.rb:8]: Yeah, it's a bit tricky. The `balance_value` is in a SELECT statement and Rails makes all columns returned by the database available as methods.

Here is the sql for the select:
https://github.com/openfoodfoundation/openfoodnetwork/blob/99378e6407a285ed9bd2ec4ec47f2e057ff71f9b/app/services/customers_with_balance.rb#L23-L29

So that serializer can only be used when you used CustomersWithBalance to query the database with the balance_value.

Comment (sauloperez) [reply app/serializers/api/admin/customer_with_balance_serializer.rb:8]: I'll work on this in a separate PR while this is being tested. I don't want this branch to live much longer and risk ending up with merge conflicts.

**Comment by filipefurtad0** [comment]
Hey @sauloperez ,

Being able to add users as beta testing as introduced by #6516 is awesome! :tada: This will definitively come in handy for prod tests.

I was trying to figure out how to enable this, for a given server/user. I'm not sure I understood correctly, but shouldn't this PR be merged first, so that the beta-users code is set up on deploy?
https://github.com/openfoodfoundation/ofn-install/pull/684/

In any case, I guess it should be possible to manually add the line `BETA_TESTERS: <a_given_user_email>` to `application.yml` and restart unicorn, as a manual workaround. But I guess we need to have it merged for deploying in production next week, right?

Anyway, I feel I need to be sure on this one, as this PR affects mostly performance, which is not easy to spot manually, in staging.

I start testing anyway -> going for the manual workaround for now :+1

**Comment by sauloperez** [comment]
All yours @filipefurtad0 . Katuma staging is provisioned and this PR now contains the changes required to use `beta_testers`.

**Comment by filipefurtad0** [comment]
The build is failing for `/spec/controllers/admin/customers_controller_spec.rb:36`, after this commit was introduced https://github.com/openfoodfoundation/openfoodnetwork/commit/f39f24e147d57864258d7b98e296179045c04c7b:

It seems related to customer balance, should this be looked at before testing or is it good to stage? ping @sauloperez

**Comment by sauloperez** [comment]
I definitely need to look at it :see_no_evil:

**Comment by sauloperez** [comment]
This was lacking a rebase to get the various beta_testers pieces that got merged. https://github.com/openfoodfoundation/openfoodnetwork/pull/6366/commits/33c72ecab021ddd74cf1896b80d2cf75ba7b627b should fix it.

**Review by Matt-Yorkley — APPROVED** [review]

File: app/serializers/api/admin/customer_with_balance_serializer.rb
Comment (sauloperez) [reply app/serializers/api/admin/customer_with_balance_serializer.rb:8]: Documented in https://github.com/openfoodfoundation/openfoodnetwork/pull/6642

**Comment by filipefurtad0** [comment]
Allright @sauloperez !

As discussed and as mentioned on the testing notes, I tested this in two steps.

Step 1) testing feature is toggle 

Significantly better performance in the `admin/customers` page will indicate that the feature is toggled. 

To verify this, I logged onto the server and tailed the log file by running `tail -f log/staging.log | grep 'ActiveRecord'` - this seems like a reasonable metric of performance. Here are some values, for a Hub with 71 customers (Tamar Valley Food Hub Wholesale). The first output line relates to the page request and the second line relates actual rendering of the customers and their balance (these are separated steps, as superadim)

a) **Before staging and toggling the feature** (this was ran several times, so these can be seen as average or representative values):

[2021-01-12 13:56:36#30737]  INFO -- : Completed 200 OK in 267ms (Views: 212.3ms | ActiveRecord: 23.4ms)
[2021-01-12 13:56:59#30737]  INFO -- : Completed 200 OK in 615ms (Views: 445.8ms | ActiveRecord: 147.3ms)

b) **After staging the PR (feature not toggled)** 

[2021-01-12 14:47:34#15577]  INFO -- : Completed 200 OK in 240ms (Views: 183.3ms | ActiveRecord: 26.0ms)
[2021-01-12 14:47:47#15577]  INFO -- : Completed 200 OK in 554ms (Views: 386.8ms | ActiveRecord: 148.1ms)

c) **After staging the PR and toggling the feature - by adding the user as a beta tester** 

[2021-01-12 15:03:12#19935]  INFO -- : Completed 200 OK in 270ms (Views: 211.8ms | ActiveRecord: 25.1ms)
[2021-01-12 15:03:22#19935]  INFO -- : Completed 200 OK in 361ms (Views: 271.2ms | ActiveRecord: 51.8ms)

The conclusion is: toggling works!! And it works only after staging and after enabling. This is the expected result, so we are good here - now we can proceed with the test cases you point out, namely, all that concerns customer balance - Step 2) in the next comment.

**Comment by filipefurtad0** [comment]
**Step 2)** Now, to the actual test cases. I added another user as a beta tester, a hub manager - and placed orders this hub as both logged in and as a guest customer. Customer balance in the `/admin/customers` was compared with `/account#/transactions#<hub_name>`, while different adjustments / order edits on made on the placed orders.

i) In all cases, it was verified that incomplete orders resulted in customer balance = 0; the customer is added to the `/admin/customers` page once the order is in the cart state, but a balance can only differ from zero if the order is complete

ii) after placing an order with cash:
- the "balance due" corresponded to the expected value
- after payment capture, the balance was = 0

iii) after placing an order with StripeSCA/Paypal:
- payment capture is immediate, so the  balance was always = 0

iv) editing orders 

- partial refunds on paid orders (balance = 0): adding/removing some items, or a line item -> the order turns to the "balance due" / "credit owed", respectively.
- partially refunding cash or Stripe -> changes the balance accordingly; refunds chance the state of the order on the customer side: an order which is "Paid" turns into "Not Paid" (this is known and current behaviour on master - one could argue it's a bug, since a partially refunded order is paid for -> I could not find an issue for this one, but I'm pretty sure I've seen it reported somewhere.
- voiding payments: voiding a payment from an order with a fully captured payment -> a paid order returns to the "balance due" state
- noticed voiding partial SCA refunds does not work - this is known #5830
- adding new payments: adding new payments on the payments section of the order
- cancelling the order: removes the balance respective of that particular order

In all the test cases i) ii) iii) and iv) the customer balance as seen from the Hub matched the values seen on the customer's account.

Reports seem to work as before - payments report `/admin/reports/payments` is blowing up, I'll need to stage master after this test, just to make sure this PR did not introduce it. Other than that, I'd say this is good to go.

**Comment by filipefurtad0** [comment]
Yup, after staging master, the report blows up as well, for dates not related with this PR. So, this looks good to me. If you feel there is that should be tested @sauloperez please let me know - moving to ready to go!

**Comment by sauloperez** [comment]
that's an amazing work @filipefurtad0 ! thanks a lot for the patience :sweat_smile: . I agree this is more than enough to ship it :ship: . What I'll do now is make sure the automated tests cover what you checked manually so that the test suite gives us the same confidence as your thorough manual tests. Don't hesitate to suggest any improvements on that regard.

**Comment by sauloperez** [comment]
> i) In all cases, it was verified that incomplete orders resulted in customer balance = 0; the customer is added to the /admin/customers page once the order is in the cart state, but a balance can only differ from zero if the order is complete

* [incomplete orders resulted in customer balance = 0](https://github.com/openfoodfoundation/openfoodnetwork/pull/6366/files#diff-fd7e9f19bdd9746f092b82080ca44424e0de0ae5b66fb46d08f7d08257f8c0feR132-R140)
* the customer is added to the /admin/customers page once the order is in the cart state
* [a balance can only differ from zero if the order is complete](https://github.com/openfoodfoundation/openfoodnetwork/pull/6366/files#diff-fd7e9f19bdd9746f092b82080ca44424e0de0ae5b66fb46d08f7d08257f8c0feR96-R109)

**Comment by sauloperez** [comment]
For the record, [this](https://app.datadoghq.com/apm/resource/rails/rack.request/65559b9abe16f5c8?durationEnd=0&durationStart=0&end=1611065298672&env=production&index=apm-search&maxPercentile=95&paused=false&query=env%3Aproduction%20service%3Arails%20operation_name%3Arack.request%20resource_name%3A%22Admin%3A%3ACustomersController%23index%22&start=1608473298672 is the) is the monitoring dashboard we should keep an eye on. Togglin it on should have a deep impact on those response times.

**Pull request merged**
